RT600 PWM support

mmahadevan108 opened the pull request:
Tested with `tests/drivers/pwm/pwm_api`. Used a scope to confirm the PWM signal.

mmahadevan108 committed:
drivers: pwm: Add PWM support for NXP LPC devices

Add PWM support using the LPC SCTimer module

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mmahadevan108 committed:
soc: RT600: Add PWM support

Add PWM support for RT600

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mmahadevan108 committed:
tests: drivers: pwm: pwm_api: add support of RT600 EVK boards

Add support for NXP RT600 EVK board

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

galak reviewed (dismissed):
Add to `tests/drivers/build_all/pwm`

mmahadevan108 committed:
test: drivers: buildall: Update to add LPC PWM

Add support for PWM on MXRT685

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mmahadevan108 commented:
> Add to `tests/drivers/build_all/pwm`

I have updated.

mmahadevan108 commented:
Anything else needed for this PR?

mbolivar-nordic commented on dts/bindings/pwm/nxp,lpc-pwm.yaml:
```diff
@@ -0,0 +1,31 @@
+    prescaler:
+      type: int
+      required: false
+      default: 1
```
All uses of default must be justified in the description as documented in https://docs.zephyrproject.org/latest/guides/dts/bindings.html#default-values-for-properties.

mmahadevan108 replied:
Thank you, I have updated the property description

mbolivar-nordic commented on tests/drivers/pwm/pwm_api/src/test_pwm.c:
```diff
@@ -185,4 +187,5 @@ void test_pwm_cycle(void)
 	/* Period : Pulse (64000 : 0), unit (cycle). Voltage : 0V */
 	zassert_true(test_task(DEFAULT_PWM_PORT, DEFAULT_PERIOD_CYCLE,
 				0, UNIT_CYCLES) == TC_PASS, NULL);
+
```
Whitespace nit

mmahadevan108 replied:
Thank you for spotting this. I removed this change.

mbolivar-nordic commented on west.yml:
```diff
@@ -100,7 +100,7 @@ manifest:
       path: tools/net-tools
     - name: hal_nxp
-      revision: 0d11138724959e1162777d9206f841ccdf64348e
+      revision: pull/90/head
```
Any progress on getting this merged?

mmahadevan108 replied:
I have pinged the team for their review comments, will update this once the PR is merged.

MaureenHelm replied:
Done

mmahadevan108 replied:
This has been updated.

mbolivar-nordic reviewed (changes requested):
> Anything else needed for this PR?

- Needs a rebase
- The HAL update needs to be merged and west.yml updated with the final SHA
- The devicetree binding uses `default` without justification

MaureenHelm commented on boards/arm/mimxrt685_evk/doc/index.rst:
```diff
@@ -129,12 +131,12 @@ functionality of a pin.
++---------+-----------------+----------------------------+
 | PIO0_29 | USART           | USART TX                   |
 +---------+-----------------+----------------------------+
 | PIO0_30 | USART           | USART RX                   |
 +---------+-----------------+----------------------------+
-| PIO0_9  | I2S             | I2S DATAIN                 |
-+---------+-----------------+----------------------------+
```
This change is not related to PWM

mmahadevan108 replied:
For some reason the I2S related items were not grouped together. I made this change while adding the PWM information.

MaureenHelm replied:
It's ok to move the I2S items together, just do it in a separate commit

mmahadevan108 replied:
Done, PR https://github.com/zephyrproject-rtos/zephyr/pull/36809 created

MaureenHelm commented on boards/arm/mimxrt685_evk/pinmux.c:
```diff
@@ -444,7 +444,7 @@ static int mimxrt685_evk_pinmux_init(const struct device *dev)
 
 #if DT_NODE_HAS_STATUS(DT_NODELABEL(flexspi), okay) && CONFIG_FLASH
-const uint32_t port1_pin11_config = (/* Pin is configured as FLEXSPI0B_DATA0 */
+	uint32_t port1_pin11_config = (/* Pin is configured as FLEXSPI0B_DATA0 */
```
Please move all the whitespace fixes into a separate commit

mmahadevan108 replied:
I will make a separate PR

mmahadevan108 replied:
PR https://github.com/zephyrproject-rtos/zephyr/pull/36785 created to address the other issues with this file.

mmahadevan108 committed:
boards: RT600: Add PWM support

Add PWM support for RT600

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

MaureenHelm reviewed (changes requested):
Can we rename this driver from `mcux_lpc` to `mcux_sctimer`? This better reflects the peripheral hardware IP and leaves room for the possibility of another PWM driver in the LPC family. I've tried to limit use of `mcux_lpc` to only cases where there are hardware IPs with the same name in different SoC families.

galak commented on dts/bindings/pwm/nxp,lpc-pwm.yaml:
```diff
@@ -0,0 +1,31 @@
+
+description: NXP LPC PWM
+
+compatible: "nxp,lpc-pwm"
```
why "nxp,lpc-pwm".  seems like we should have sctimer in the name.

mmahadevan108 replied:
I have renamed the driver to `pwm_mcux_sctimer.c` and changed this to `nxp_sctimer_pwm`

mmahadevan108 commented:
> Can we rename this driver from `mcux_lpc` to `mcux_sctimer`? This better reflects the peripheral hardware IP and leaves room for the possibility of another PWM driver in the LPC family. I've tried to limit use of `mcux_lpc` to only cases where there are hardware IPs with the same name in different SoC families.

I have renamed the driver to `pwm_mcux_sctimer.c` and have renamed the configs, device tree related items and the function/variable names in the driver.

MaureenHelm commented on boards/arm/mimxrt685_evk/mimxrt685_evk_cm33.dts:
```diff
@@ -19,6 +19,8 @@
 		led2 = &red_led;
 		usart-0 = &flexcomm0;
+		/* For pwm test suites */
+		pwm-0 = &sc_timer;
```
Can the LEDs on this board be driven by a PWM? If so, please add pwm-led aliases to enable `samples/basic/rgb_led` and `samples/drivers/led_pwm`

mbolivar-nordic reviewed (approved):
my comments are addressed

LeBlue commented:
There is a major issue with the current implementation of this driver. The peripheral has a limitation that is also documented in the hal API that is used to set/change the PWM period:
[SCTIMER_SetupPwm](https://github.com/zephyrproject-rtos/hal_nxp/blob/6d34c83c485865c25f90456d00e054947ca54978/mcux/drivers/imxrt6xx/fsl_sctimer.c#L240)

In short: the PWM period is not (really) reconfigurable (hal API/implementation limitation) and shared between all channels (peripheral limitation), only the pulse width is changeble per channel. Changing it for one channel changes the period for all. The driver does not prevent that, but it should.

I have tested it with the lpcxresso55s28 board (should be the same IP core) and the blinky_pwm example that fails because the limitation (driver errors out after changing period for more than numer of available channels.)

To make the driver usable for now with the limitation and minimal changes, the period could be fixed after first configuration.

MaureenHelm committed:
boards: arm: Add PWM LEDs to the mimxrt685_evk board

Adds support for PWM LEDs (red, green, blue) to the mimxrt685_evk board
by adding devicetree nodes and aliases, and configuring the associated
pinmuxes. The red PWM LED is disabled by default because it's connected
to the same PWM channel as the blue PWM LED.

Signed-off-by: Maureen Helm <maureen.helm@nxp.com>

MaureenHelm commented:
- Rebased and resolved merge conflicts
- Added a new commit to enable PWM LEDs on the board. I tested with `samples/basic/fade_led` and found an issue when the pulse width is set zero. I'm not sure what the right resolution is yet, but the sample works correctly if I comment out the line that restarts the timer in the 0% duty cycle case.

> In short: the PWM period is not (really) reconfigurable (hal API/implementation limitation) and shared between all channels (peripheral limitation), only the pulse width is changeble per channel. Changing it for one channel changes the period for all. The driver does not prevent that, but it should.

It's not unusual for all channels to share the same period, some drivers log a warning: https://github.com/NXPmicro/zephyr/blob/b8820da4e6341ff4fe89fc7e1a23c1995ece1b43/drivers/pwm/pwm_mcux_ftm.c#L111-L114

> I have tested it with the lpcxresso55s28 board (should be the same IP core) and the blinky_pwm example that fails because the limitation (driver errors out after changing period for more than numer of available channels.)

Thanks for testing. I see the same behavior on the `mimxrt685_evk_cm33` board.

mmahadevan108 commented:
Thank you. I have added a fix to the driver for the case when the dutycyle is zero to check if the PWM has been setup before starting the timer.
